File: Tests/DungeonSoundboardCoreTests/ThemeRendererTests.swift
import XCTest
@testable import DungeonSoundboardCore

final class ThemeRendererTests: XCTestCase {
    func testThemeRoundTripPreservesValues() throws {
        let original = ThemeRenderer.theme(for: .forestMist)
        let data = try JSONEncoder().encode(original)
        let decoded = try JSONDecoder().decode(AppTheme.self, from: data)

        XCTAssertEqual(decoded, original)
    }

    func testThemeDecodeFallsBackForMissingFields() throws {
        let json = #"{"preset":"classicDungeon","palette":{"accent":{"red":0.2,"green":0.4,"blue":0.6,"alpha":1}}}"#
        let decoded = try JSONDecoder().decode(AppTheme.self, from: Data(json.utf8))

        XCTAssertEqual(decoded.preset, .classicDungeon)
        XCTAssertEqual(decoded.palette.accent, ThemeColor(red: 0.2, green: 0.4, blue: 0.6))
        XCTAssertEqual(decoded.palette.backgroundTop, ThemeRenderer.defaultTheme.palette.backgroundTop)
        XCTAssertEqual(decoded.chrome.cornerRadius, ThemeRenderer.defaultTheme.chrome.cornerRadius)
    }

    func testPresetApplicationProvidesExpectedIdentity() {
        let tavern = ThemeRenderer.theme(for: .tavernEmber)

        XCTAssertEqual(tavern.preset, .tavernEmber)
        XCTAssertEqual(tavern.chrome.panelBlurStrength, .medium)
        XCTAssertEqual(tavern.chrome.density, .normal)
    }

    func testSanitizeClampsUnsafeNumericValues() {
        var theme = ThemeRenderer.defaultTheme
        theme.background.opacity = 99
        theme.background.dimOverlay = -10
        theme.background.blurRadius = 100
        theme.chrome.accentIntensity = -4
        theme.chrome.panelOpacity = 0.1
        theme.chrome.cornerRadius = 50

        let sanitized = ThemeRenderer.sanitize(theme)

        XCTAssertEqual(sanitized.background.opacity, 1)
        XCTAssertEqual(sanitized.background.dimOverlay, 0.15)
        XCTAssertEqual(sanitized.background.blurRadius, 24)
        XCTAssertEqual(sanitized.chrome.accentIntensity, 0)
        XCTAssertEqual(sanitized.chrome.panelOpacity, 0.55)
        XCTAssertEqual(sanitized.chrome.cornerRadius, 24)
    }

    func testReadabilityGuardrailsRaiseContrast() {
        let surface = ThemeColor(red: 0.15, green: 0.15, blue: 0.15)
        let text = ThemeColor(red: 0.16, green: 0.16, blue: 0.16)

        let improved = ThemeRenderer.makeReadable(text, against: surface, minimumContrast: 4.5)

        XCTAssertGreaterThanOrEqual(improved.contrastRatio(against: surface), 4.5)
    }

    func testBackgroundConfigSurvivesCoding() throws {
        let config = BackgroundConfig(
            mode: .image,
            imageBookmarkData: Data([0x01, 0x02]),
            imageOriginalPath: "/tmp/bg.png",
            layoutMode: .tile,
            opacity: 0.5,
            dimOverlay: 0.4,
            blurRadius: 6
        )

        let data = try JSONEncoder().encode(config)
        let decoded = try JSONDecoder().decode(BackgroundConfig.self, from: data)

        XCTAssertEqual(decoded, config)
    }
}

File: windows/tests/DungeonSoundboard.Core.Tests/ThemeRendererTests.cs
using System.Text.Json;
using DungeonSoundboard.Core.Models;
using DungeonSoundboard.Core.Serialization;
using Xunit;

namespace DungeonSoundboard.Core.Tests;

public sealed class ThemeRendererTests
{
    [Fact]
    public void ThemeRoundTripPreservesValues()
    {
        var original = ThemeRenderer.ThemeFor(ThemePreset.ForestMist);
        var data = JsonSerializer.SerializeToUtf8Bytes(original, JsonDefaults.Options);
        var decoded = JsonSerializer.Deserialize<AppTheme>(data, JsonDefaults.Options);

        Assert.Equal(original, decoded);
    }

    [Fact]
    public void ThemeDecodeFallsBackForMissingFields()
    {
        const string json = """
        {"preset":"classicDungeon","palette":{"accent":{"red":0.2,"green":0.4,"blue":0.6,"alpha":1}}}
        """;

        var decoded = JsonSerializer.Deserialize<AppTheme>(json, JsonDefaults.Options);

        Assert.NotNull(decoded);
        Assert.Equal(ThemePreset.ClassicDungeon, decoded.Preset);
        Assert.Equal(new ThemeColor(0.2, 0.4, 0.6), decoded.Palette.Accent);
        Assert.Equal(ThemeRenderer.DefaultTheme.Palette.BackgroundTop, decoded.Palette.BackgroundTop);
        Assert.Equal(ThemeRenderer.DefaultTheme.Chrome.CornerRadius, decoded.Chrome.CornerRadius);
    }

    [Fact]
    public void PresetApplicationProvidesExpectedIdentity()
    {
        var tavern = ThemeRenderer.ThemeFor(ThemePreset.TavernEmber);

        Assert.Equal(ThemePreset.TavernEmber, tavern.Preset);
        Assert.Equal(PanelBlurStrength.Medium, tavern.Chrome.PanelBlurStrength);
        Assert.Equal(InterfaceDensity.Normal, tavern.Chrome.Density);
    }

    [Fact]
    public void SanitizeClampsUnsafeNumericValues()
    {
        var theme = ThemeRenderer.DefaultTheme;
        theme.Background.Opacity = 99;
        theme.Background.DimOverlay = -10;
        theme.Background.BlurRadius = 100;
        theme.Chrome.AccentIntensity = -4;
        theme.Chrome.PanelOpacity = 0.1;
        theme.Chrome.CornerRadius = 50;

        var sanitized = ThemeRenderer.Sanitize(theme);

        Assert.Equal(1, sanitized.Background.Opacity);
        Assert.Equal(0.15, sanitized.Background.DimOverlay);
        Assert.Equal(24, sanitized.Background.BlurRadius);
        Assert.Equal(0, sanitized.Chrome.AccentIntensity);
        Assert.Equal(0.55, sanitized.Chrome.PanelOpacity);
        Assert.Equal(24, sanitized.Chrome.CornerRadius);
    }

    [Fact]
    public void ReadabilityGuardrailsRaiseContrast()
    {
        var surface = new ThemeColor(0.15, 0.15, 0.15);
        var text = new ThemeColor(0.16, 0.16, 0.16);

        var improved = ThemeRenderer.MakeReadable(text, surface, 4.5);

        Assert.True(improved.ContrastRatio(surface) >= 4.5);
    }

    [Fact]
    public void BackgroundConfigSurvivesCoding()
    {
        var config = new BackgroundConfig
        {
            Mode = BackgroundMode.Image,
            ImageBookmarkData = [0x01, 0x02],
            ImageOriginalPath = "C:\\images\\bg.png",
            LayoutMode = BackgroundLayoutMode.Tile,
            Opacity = 0.5,
            DimOverlay = 0.4,
            BlurRadius = 6
        };

        var data = JsonSerializer.SerializeToUtf8Bytes(config, JsonDefaults.Options);
        var decoded = JsonSerializer.Deserialize<BackgroundConfig>(data, JsonDefaults.Options);

        Assert.Equal(config, decoded);
    }
}
